osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Input;
+using osu.Framework.Input.Bindings;
+using osu.Framework.Input.Events;
+
+namespace osu.Game.Rulesets.Osu.UI
+{
+    public class OsuTouchInputMapper : Drawable
+    {
+        /// <summary>
+        /// The default cursor touch it may fail <see cref="IsCursorTouch(TouchSource)"/> on cases where there aren't any cursor touches, such as when autopilot is on.
+        /// </summary>
+        public const TouchSource DEFAULT_CURSOR_TOUCH = TouchSource.Touch1;
+
+        /// <summary>
+        /// The maximum amount of tap touches that should be allowed.
+        /// </summary>
+        private const int allowed_tap_touches_limit = 2;
+
+        /// <summary>
+        /// The maximum amount of tap touches that should be allowed decremented. this is useful for calculations that require to account for the next touch.
+        /// </summary>
+        private const int allowed_tap_touches_limit_decremented = allowed_tap_touches_limit - 1;
+
+        /// <summary>
+        /// Our parent <see cref="OsuInputManager"/>
+        /// </summary>
+        private readonly OsuInputManager osuInputManager;
+
+        /// <summary>
+        /// Our <see cref="KeyBindingContainer"/> used to trigger the touch actions from the <see cref="tapTouchActions"/>.
+        /// </summary>
+        private KeyBindingContainer<OsuAction> keyBindingContainer => osuInputManager.KeyBindingContainer;
+
+        /// <summary>
+        /// A dictionary that maps <see cref="TouchSource"/> into a respective <see cref="OsuAction"/>.
+        /// </summary>
+        private readonly Dictionary<TouchSource, OsuAction> tapTouchActions = Enum.GetValues(typeof(TouchSource))
+                                                                                  .Cast<TouchSource>()
+                                                                                  .ToDictionary(source => source, source => (source - TouchSource.Touch1) % 2 == 0 ? OsuAction.LeftButton : OsuAction.RightButton);
+
+        /// <summary>
+        /// Tracks all active tap <see cref="TouchSource"/>s that can be mapped into a given <see cref="OsuAction"/> through the <see cref="tapTouchActions"/>.
+        /// </summary>
+        private readonly HashSet<TouchSource> activeTapTouches = new HashSet<TouchSource>();
```

Comment:
Why is this required when `tapTouchActions` already exists? Seems like it's storing the same thing.

Reply:
It is required because these are the tap touch sources that were handled by the `OnTouchDown` so they can be removed and their assigned action be released later.
i though on just making a count of sctive tap touches and incrementing when a touch happens and getting the current count `OnTouchUp` and releasing the touch for the index of said count, but since tap ordering isn't guaranteed this does not work, so i believe this is a must.

Reply:
Also these active tap touches track how many tap touches are currently active so we don't accept tap input for too many fingers.